### PR TITLE
feat: pieces v2!

### PR DESCRIPTION
**Pull request by kyranet**
## Changes

- Added default loader strategy in Store.
- Added Dependency Injection support for sending context into pieces.
- Added support to dynamically load CJS and ESM.
- Added support to load all structures instead of only one.
- Make errors show their type in the name.
- Moved all separate hooks into a single LoaderStrategy.
- Renamed `Piece#extras` to `Piece#context`.
- Removed `PieceContext#extras`.

### CODE REVIEW

File: src/lib/strategies/LoaderStrategy.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		// Retrieve the name of the file, return null if empty.
+		const name = basename(path, extension);
+		if (name === '') return null;
+
+		// Return the name and extension.
+		return { extension, name };
+	}
+
+	public preload(path: string): PreloadResult<T> {
+		return import(path);
```
Comment (vladfrangu) [inline src/lib/strategies/LoaderStrategy.ts:26]: Why not do the timestamp thing too, so we don't need a custom loader to reload pieces 👀

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

File: src/lib/Store.ts [hunk 2]
```diff
@@ -340,7 +191,80 @@ export class Store<T extends Piece> extends Collection<string, T> {
...
 	}
+
+	/**
+	 * The injected variables that will be accessible to all stores and pieces. To add an extra property, simply mutate
+	 * the object to add it, and this will update all stores and pieces simultaneously.
+	 *
+	 * @example
+	 * ```typescript
+	 * // Add a reference to the Client:
+	 * import { Store } from '(at)sapphire/pieces';
```
Comment (cfanoulis) [inline src/lib/Store.ts:205]:
```suggestion
	 * import { Store } from '@sapphire/pieces';
```

File: src/lib/Store.ts [hunk 3]
```diff
@@ -340,7 +191,80 @@ export class Store<T extends Piece> extends Collection<string, T> {
...
+	 *   constructor(options) {
+	 *     super(options);
+	 *
+	 *     Store.injectedContext.client = this;
+	 *   }
+	 * }
+	 *
+	 * // Can be placed anywhere in a TypeScript file, for JavaScript projects,
+	 * // you can create an `augments.d.ts` and place the code there.
+	 * declare module '(at)sapphire/pieces' {
```
Comment (cfanoulis) [inline src/lib/Store.ts:217]:
```suggestion
	 * declare module '@sapphire/pieces' {
```
whats up with not using @ now

File: src/lib/strategies/LoaderStrategy.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import { URL } from 'url';
+import { MissingExportsError } from '../errors/MissingExportsError';
+import { mjsImport } from '../internal/import';
+import type { Piece } from '../Piece';
+import type { Store } from '../Store';
+import type { AsyncPreloadResult, FilterResult, ILoaderResult, ILoaderStrategy, ModuleData } from './ILoaderStrategy';
+import { classExtends, isClass } from './Shared';
+
+/**
+ * An abstracted loader interface.
```
Comment (cfanoulis) [inline src/lib/strategies/LoaderStrategy.ts:11]: Comment ≠ actual code

**Review by cfanoulis — COMMENTED** [review]
minor nits

File: src/lib/Store.ts (hunk 2 above)
Comment (kyranet) [reply src/lib/Store.ts:205]: That breaks VSCode's IntelliSense.

Comment (cfanoulis) [reply src/lib/Store.ts:205]: has anyone submitted a bug report lol

anyways /resolved (cant resolve on mobile)

File: src/lib/Store.ts (hunk 3 above)
Comment (cfanoulis) [reply src/lib/Store.ts:217]: /resolved see previous comment

**Review by cfanoulis — APPROVED** [review]

**Review by Favna — APPROVED** [review]

**Pull request merged**